das/daser.go

```diff
@@ -68,195 +72,82 @@ func (d *DASer) Start(context.Context) error {
-	return nil
-}
+		h, err := d.getter.Head(ctx)
+		if err == nil {
+			cp = checkpoint{
+				MinSampled: 1,
+				MaxKnown:   uint64(h.Height),
+			}
 
-// Stop stops sampling.
-func (d *DASer) Stop(ctx context.Context) error {
-	// Stop func can now be invoked twice in one Lifecycle, when:
-	// * BEFP is received;
-	// * node is stopping;
-	// this statement helps avoiding panic on the second Stop.
-	// If ctx.Err is not nil then it means that Stop was called for the first time.
-	// NOTE: we are expecting *only* ContextCancelled error here.
-	if d.ctx.Err() == context.Canceled {
-		return nil
-	}
-	d.cancel()
-	// wait for both sampling routines to exit
-	for i := 0; i < 2; i++ {
-		select {
-		case <-d.catchUpDn:
-		case <-d.sampleDn:
-		case <-ctx.Done():
-			return ctx.Err()
+			log.Warnw("set max known height from store", "height", cp.MaxKnown)
 		}
 	}
-	d.cancel = nil
-	return nil
-}
-
-// sample validates availability for each Header received from header subscription.
-func (d *DASer) sample(ctx context.Context, sub header.Subscription, checkpoint int64) {
-	// indicate sampling routine is running
-	d.indicateRunning()
-	defer func() {
-		sub.Cancel()
-		// indicate sampling routine is stopped
-		d.indicateStopped()
-		// send done signal
-		d.sampleDn <- struct{}{}
-	}()
-
-	// sampleHeight tracks the last successful height of this routine
-	sampleHeight := checkpoint
-
-	for {
-		h, err := sub.NextHeader(ctx)
-		if err != nil {
-			if err == context.Canceled {
-				return
-			}
+	log.Infow("loaded checkpoint", "height", cp)
 
-			log.Errorw("failed to get next header", "err", err)
-			continue
-		}
+	d.sampler.state = initSamplingState(samplingRange, cp)
 
-		// If the next header coming through gossipsub is not adjacent
-		// to our last DASed header, kick off routine to DAS all headers
-		// between last DASed header and h. This situation could occur
-		// either on start or due to network latency/disconnection.
-		if h.Height > sampleHeight+1 {
-			// DAS headers between last DASed height up to the current
-			// header
-			job := &catchUpJob{
-				from: sampleHeight,
-				to:   h.Height - 1,
-			}
-			select {
-			case <-ctx.Done():
-				return
-			case d.jobsCh <- job:
-			}
-		}
+	runCtx, cancel := context.WithCancel(context.Background())
+	d.cancel = cancel
 
-		err = d.sampleHeader(ctx, h)
-		if err != nil {
-			// record error
-			d.updateSampleState(h, err)
-			log.Warn("DASer SAMPLING ROUTINE WILL BE STOPPED. IN ORDER TO CONTINUE SAMPLING, " +
-				"RE-START THE NODE")
-			return
-		}
+	go d.sampler.runCoordinator(runCtx)
+	// subscribe to new headers
+	go d.runSubscriber(runCtx, sub, d.sampler.listen)
 
-		d.updateSampleState(h, nil)
-		sampleHeight = h.Height
-	}
+	return nil
 }
 
-// catchUpJob represents a catch-up job. (from:to]
-type catchUpJob struct {
-	from, to int64
-}
+// Stop stops sampling.
+func (d *DASer) Stop(ctx context.Context) error {
+	if atomic.CompareAndSwapInt32(&d.closed, 0, 1) {
```

Comment:
stylistic nit: let's keep logic outside of the if case and keep return nil there instead